### ISSUE TITLE
Data NFT

### ISSUE THREAD

**Issue by kremalicious** [oceanprotocol/market#944]
Our [NFT factory](https://github.com/oceanprotocol/contracts/blob/v4main/contracts/ERC721Factory.sol) allows to have the following metadata about the token:

- name
- symbol
- tokenURI

For tokenURI we mainly want to use that to embed a unique image for each NFT among other metadata about the token. Now, having a hard time finding some standard around the usage of `tokenURI` or `tokenMetadata` in other 3rd party contracts where it's not exactly clear how e.g. an image has to be embedded/referenced for majority of NFT wallets to pick this up out of the box.

## Main goals

- create implementation concept for creating new Ocean v4 data NFTs: name? symbol? image?
- most decentralized image & token metadata, ideally no external image hosting whatsoever
- most future-proof image & token metadata, ideally not even IPFS gateway URLs

## Currently implemented flow

PoC over in https://github.com/oceanprotocol/market/pull/933 follows how Uniswap v3 does it with their liquidity position NFTs where the `tokenURI` resolves to an object after base64 decoding and JSON parsing it. The SVG image is embedded in there as `data-uri`, allowing image display without external hosting. We like that.

1. generate a SVG in code
2. URL encode SVG code string, and make data-uri out of it
3. embed in json object under `image` key
4. base64 encode the json object, and put onto final `tokenURI`
5. when displaying NFT: decode `tokenURI`, parse as json, use `image` key value as `src` for an `img`.

 So in our case the base64 encoded json resolves to:

```json
{
  "name": "Ocean Asset v4 NFT",
  "symbol": "OCEAN-V4-NFT",
  "description": "This NFT represents an asset in the Ocean Protocol v4 ecosystem.",
  "image": "data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='M0 10.4304L16.3396 10.4304L8.88727 17.6833L10.2401 19L20 9.5L10.2401 0L8.88727 1.31491L16.3396 8.56959L0 8.56959V10.4304Z' /%3E%3C/svg%3E"
}
```

This json schema follows the [OpenSea metadata standard](https://docs.opensea.io/docs/metadata-standards), we only need to figure out if embedded SVG works there too, or if we have to use `ipfs://xxxx` for some referenced image.

Currently all abstracted into `@utils/nft`: https://github.com/oceanprotocol/market/blob/v4-publish/src/%40utils/nft.ts

Which, with the arrow test image, ends up like this in current UI draft:

<img width="900" alt="Screen Shot 2021-11-16 at 10 27 53" src="https://user-images.githubusercontent.com/90316/141968591-ace0674c-8a9c-48d2-8193-b5cffcb07266.png">

## Name & Symbol

Still not sure if we should introduce some wordlist-generated names like we do with datatokens. If we leave the maritime names to datatokens only it is actually a nice distinction between NFT & datatokens. For the NFT we could simply go with 

- name: `Ocean Asset v4 NFT`
- symbol: `OCEAN-V4-NFT`

Name should be only important to get this as the collection name on OpenSea, so we need to check if that already works in current implementation. Then they all should have the same name.

For all of them  as the contract address distinguish them, technically. But having some unique reference in there ideally connecting it to the dataset would be great though. Some candidates to store in `name` & `description` in the json:

- DID
- NFT contract address
- asset URL in Ocean Market

These all need to be checked when exactly in NFT creation process we have which information.

## Visuals

This all just touches the technical part and already keeps us busy. As we do not have the resources to now create a nice generative art library, we might want to go simple and figure out how to use our code for the waves on oceanprotocol.com to render custom static images, as those waves are already live-generated SVG so should be adaptable.

Update: And just like that over in #933, preview: https://market-git-v4-publish-oceanprotocol.vercel.app/publish 
<img width="894" alt="Screen Shot 2021-11-16 at 21 15 02" src="https://user-images.githubusercontent.com/90316/142066864-c0a94922-d02a-4182-a445-873ff729be29.png">

## 3rd party reverse engineering

Update: pretty much what the [OpenSea metadata standards](https://docs.opensea.io/docs/metadata-standards) say. Worth noting that in my tests using `"image": "ipfs://xxx"` works beautifully on OpenSea for a referenced image so at least we do not have to use some IPFS gateway URL.

## Possible problems

Wallets would then have to follow the assumption to decode the `tokenURI`, parse it as JSON, and find an `image` key. Not sure they all do, I would assume most then have custom integrations e.g. for Uniswap or Rarible. So we could end up chasing wallet developers like we had to do when integrating OCEAN token properly in most wallets.

Embedding the SVG directly might blow up data size for what's stored on-chain, resulting in increased transaction costs. Could be more mitigated with further SVG optimizations.

## Solution

We gotta try and test. Create NFT with currently implemented flow, then checkout publisher account in some widely used wallets/sites:

- [Rainbow Wallet](https://rainbow.me)
- [MetaMask mobile](https://metamask.io)
- [Rarible](https://rarible.com)
- [OpenSea](https://opensea.io)
- [Zerion](https://zerion.io)
- [Zapper](https://zapper.fi)

Out of this list, only Rainbow & MetaMask allow to switch to Rinkeby. Finally we also need to document all this in https://github.com/oceanprotocol/docs/pull/780 which currently does not even mention `tokenURI` or anything really useful for implementation.

If we need to host an image or some json file somewhere instead of embedding it, then only IPFS would be candidate but we most likely then don't get around storing a gateway URL instead of IPFS content hash as wallets/sites most likely know what to do with URL but not with content hash.

On top of that we need to gather as much information as possible about 3rd party implementations of NFT display, and get to know their assumptions.

**Comment by kremalicious** [comment]
quick test, in Rainbow wallet the Uniswap positions do net get an image but they still somehow read out the `description` value from the base64 json behind `tokenURI`. Or might be Uniswap changed something and the very first NFTs created with v3 do not show anymore:

<img src="https://user-images.githubusercontent.com/90316/141976184-304ccda8-ed96-4f72-93b6-3fdeadecf88e.PNG" width="300" />

**Comment by kremalicious** [comment]
Next steps:

- go with what we have in #933 and see how that goes with first publish tests
- but the generated SVG needs to be tweaked further as the generated waves are actually way wider than what's visible, creating unnecessary large SVG. We need to find a way to crop the generated SVG to the view we are presenting in the form
- all the `TODO` in `@utils/nft` can be worked on but they are dependent on actual publishing and checking out created NFT in some wallet

**Comment by mihaisc** [comment]
Tested with the following in tokenURI : svg ,  json with metadata, url to metadata json (like opensea, some with exactly the same link). For some reason none of them work , the image is not visible on opensea or metamask mobile wallet ( metamask extension doesn't support nft)
![image](https://user-images.githubusercontent.com/26000280/146956407-839c2872-8c2b-4eea-aca6-533a18b64421.png)

**Comment by kremalicious** [comment]
updated my initial description a bit, basically following the OpenSea metadata standards should give us best chance of many wallets supporting those: https://docs.opensea.io/docs/metadata-standards#metadata-structure

Current route is to rather use the `image_data` key in the json file for the embedded SVG and see if that works. But also looks like our current test NFTs all have an empty `tokenURI` which would be a new issue.

**Comment by mihaisc** [comment]
I suggest shortening the name to `OCEAN-NFT` (there are no v3 nfts anyway, the v4 part is redundant) so that we can add them to metamask , the current name has one character over the limit. In the future metamask (browser extension) will have nft support but we don't know when that will be.
![image](https://user-images.githubusercontent.com/26000280/149314862-d3ac0dd8-645d-4f8e-8fca-656b16803107.png)

**Comment by trentmc** [comment]
> I suggest shortening the name to OCEAN-NFT 

Good suggestion. I'm ok with this.

**Comment by oceanByte** [comment]
> Did some test with the new nft, image not visible in metamask mobile. Did you manage to make the image visible in any wallet? 

The issue is still pending, as discussed in the sprint plannings, as we depend on feedback from Etherscan here. We could not find any mistake in the contracts (e.g., missing events, not following ERC721 standard, wrong method names, etc.). Some big NFT projects have indeed a custom implementation on NFT platforms and Etherscan to be displayed correctly.

We do not have a direct contact at Etherscan, so we had to go through the official route via https://etherscan.io/contactus. But there was still no answer yet. We will send a new request.

Our implemented solution (https://github.com/oceanprotocol/market/pull/1032) is based on our successful test using base64 encoded JSON inside an ERC721 contract with an SVG as image_data. It works for OpenSea and rainbow wallet: https://testnets.opensea.io/assets/0xb017a2698db9fb53142f5c342a6533a8be38d1a2/1

I will compile a detailed status update during the weekend to provide insights into our approaches to keep it documented here.

**Comment by oceanByte** [comment]
**Current status**: We hoped to get feedback by now and provide a solution but we are still awaiting feedback from the Etherscan Support Center to understand why the number of holders is not updated. We assume that this issue is related to other problems such as not properly displaying the NFT on platforms like OpenSea or wallets like Rainbow Wallet.

## Etherscan Issues
There are two main problems with the current implementation of the contracts: 1) Etherscan does not show the number of holders nor transfers (always zero) and 2) while it shows an ERC-721 transfer in the detail view, ERC-20 is shown when using the token tracker. 

Transaction Details: https://rinkeby.etherscan.io/tx/0xbdc4b2eebbf23f4f7e99877e6f5e93272cc1115212488b68c7c7fba061cad1f5
<img width="515" alt="image" src="https://user-images.githubusercontent.com/75223088/157630559-2ebee7ad-bb1e-45aa-baf4-d5fecd7052e8.png">

Token Tracker: https://rinkeby.etherscan.io/token/0x690fc8d36beb4af84b5602baf33f64cbd4303bff#readContract
<img width="597" alt="image" src="https://user-images.githubusercontent.com/75223088/157630508-cd731961-2f8a-48c9-8010-8728b8ccbaf0.png">

**Usage of Factory**. We assumed that this might be related to the factory that we are using. Other projects are also using factories and Etherscan picks it up correctly. See https://etherscan.io/token/0x7C770595a2Be9A87CF49B35eA9bC534f1a59552D. And here is also a detailed transaction that uses several other contracts before minting ERC-721 tokens: https://etherscan.io/tx/0x5bcef80c936ae6c00b5d71aa7dcf1b0e42fd5bfcc34c3ffd726d32d74b342b97

We could not find any problem with the OPF implementation, but it might be worth investing more time here. As deploying NFT contracts without the factory works as intended. 

**Tracking ERC-721 non-fungible token on Etherscan**. As not all platforms handle NFTs in the same way, we looked for any Etherscan specific requirements. We found an old blog entry by Etherscan on how they track ERC-721 tokens here: https://medium.com/etherscan-blog/tracking-erc-721-non-fungible-token-on-etherscan-b4da1ed955ed 

> "The token contract must emit at least one ERC-721 compliant transfer event which looks something like"
> 
> `event Transfer(address indexed _from, address indexed _to, uint256 indexed _tokenId);`

The Transfer event is emitted successfully when the token is created. See: https://rinkeby.etherscan.io/address/0x690fc8d36beb4af84b5602baf33f64cbd4303bff#events

It also states that images must be added manually to Etherscan, but the post is quite old (2018). So this might have changed. There is actually an NFT page available on Etherscan but this feature is not available for testnet ("Sorry! This feature is not available in Testnet"):

https://etherscan.io/nft/0x4a537F61ef574153664c0Dbc8c8F4B900cacBE5d/1
<img width="598" alt="image" src="https://user-images.githubusercontent.com/75223088/157631183-67af2d87-fb06-4b2e-b963-0ab823509a29.png">


## Displaying in Wallets and platforms

**MetaMask (Desktop)** recognizes the given contract and automatically selects the name (OCEAN-V4-NFT) and decimal (0). But the amount is zero.

<img width="249" alt="image" src="https://user-images.githubusercontent.com/75223088/157631232-d2a9784a-2f48-4eea-8180-d2de019be776.png">

**OpenSea**
We also tested everything in isolation. And figured out that OpenSea requires ERC-165 to be implemented while Etherscan has a bug where the detail page shows an ERC-721 transfer but the Token Tracker shows an ERC-20 token: 

With ERC-165:  [OpenSea](https://testnets.opensea.io/assets/0x8b5af742c8bcf90b25392a425a41cc7b57c0c459/1) | [Etherscan](https://rinkeby.etherscan.io/tx/0xf01289712e16332dbbcaea05865b2a06e0c1acac1bc97db27026daa17eb288fe)

Without ERC-165: [OpenSea](https://testnets.opensea.io/assets/0x04f79316c0bb41199bc51a70ae7b06beaa017464/1) | [Etherscan](https://rinkeby.etherscan.io/tx/0x0ff55fa803967352ecca46e6d4343f5871327093dee0e83ae0ff824cc982896b)

OpenSea has an API to test contracts. When providing it an Ocean v4 NFT contract it fails to validate it properly:
https://testnets-api.opensea.io/api/v1/asset/0xf685147c14f15743b623530c5e31c61558122c67/1/validate
We can force a validation though which returns `success: false`: 
https://testnets-api.opensea.io/api/v1/asset/0xf685147c14f15743b623530c5e31c61558122c67/1/?force_update=true&format=json

Apart from that, we know that base64 encoded JSON inside of contract with an SVG as image_data works as intended. See here: https://testnets.opensea.io/assets/0xb017a2698db9fb53142f5c342a6533a8be38d1a2/1 This also shows up correctly in the Rainbow wallet. And it is successfully validated by OpenSea: https://testnets-api.opensea.io/api/v1/asset/0xb017a2698db9fb53142f5c342a6533a8be38d1a2/1/validate

**Comment by alexcos20** [comment]
fixed.

See: https://rinkeby.etherscan.io/token/0x53c9865977b0003f39204a7943a1d96571a224d0#balances

and commit on contracts: https://github.com/oceanprotocol/contracts/pull/556/commits/009db99f1c97ea73b0ea553589d9000dc31dc7d9

Market will have to wait for a ocean-contracts release , then ocean-js release.   Will resume once this is done
